renderdoc/driver/ihv/nv/nv_counters.cpp

```diff
@@ -0,0 +1,380 @@
+{
+}
+
+NVCounters::~NVCounters()
+{
+  if(mObjectsCount != 0)
+  {
+    NvPmResultFails(mNvPmApi->DeleteObjects(mNvPmCtx), "call to 'NvPmApi::DeleteObjects'");
+    mObjectsCount = 0;
+  }
+
+  if(mNvPmCtx != static_cast<uint64_t>(-1))
+  {
+    NvPmResultFails(mNvPmApi->DestroyContext(mNvPmCtx), "call to 'NvPmApi::DestroyContext'");
+    mNvPmCtx = static_cast<uint64_t>(-1);
+  }
+
+  if(mNvPmApi != 0)
+  {
+    NvPmResultFails(mNvPmApi->Shutdown(), "call to 'NvPmApi::Shutdown'");
+    mNvPmApi = NULL;
+  }
+  mNvPmLib = NULL;
+}
+
+bool NVCounters::Init()
+{
+  if(mNvPmLib != NULL)
+  {
+    return false;
+  }
+
+  mNvPmLib = Process::LoadModule("NvPmApi.Core.dll");
+  if(mNvPmLib == NULL)
+  {
+    RDCWARN("NV GPU performance counters could not locate 'NvPmApi.Core.dll'");
+    return false;
+  }
+
+  NVPMGetExportTable_Pfn pfnGetExportTable =
+      (NVPMGetExportTable_Pfn)Process::GetFunctionAddress(mNvPmLib, "NVPMGetExportTable");
+  if(pfnGetExportTable == NULL)
+  {
+    return false;
+  }
+
+  if(NvPmResultFails(pfnGetExportTable(&ETID_NvPmApi, (void **)&mNvPmApi), "get 'NvPmApi' table"))
+  {
+    return false;
+  }
+
+  if(NvPmResultFails(mNvPmApi->Init(), "init 'NvPmApi'"))
+  {
+    return false;
+  }
+  return true;
+}
+
+bool NVCounters::Init(ID3D11Device *pDevice)
+{
+  if(Init() == false)
+  {
+    return false;
+  }
+
+  if(NvPmResultFails(mNvPmApi->CreateContextFromD3D11Device(pDevice, &mNvPmCtx),
+                     "init 'NVPMContext' from ID3D11Device"))
+  {
+    return false;
+  }
+
+  uint32_t NumCounters = 0;
+  mNvPmApi->EnumCountersByContextUserData(mNvPmCtx, NvPmCountCounters, &NumCounters);
+
+  EnumCountersCtx ctx;
+  ctx.mExternalIds.resize(NumCounters);
+  ctx.mInternalIds.resize(NumCounters);
+  ctx.mExternalDescriptors.resize(NumCounters);
+  ctx.mInternalDescriptors.resize(NumCounters);
+  ctx.mNvPmApi = mNvPmApi;
+  ctx.mCurrentCounterId = 0;
+
+  mNvPmApi->EnumCountersByContextUserData(mNvPmCtx, NvPmGatherCounters, &ctx);
+
+  ctx.mExternalIds.swap(mExternalIds);
+  ctx.mInternalIds.swap(mInternalIds);
+  ctx.mExternalDescriptors.swap(mExternalDescriptors);
+  ctx.mInternalDescriptors.swap(mInternalDescriptors);
+
+  mSelectedExternalIds.reserve(NumCounters);
+  mSelectedInternalIds.reserve(NumCounters);
+
+  return true;
+}
+
+bool NVCounters::PrepareExperiment(const std::vector<GPUCounter> &counters, uint32_t objectsCount)
+{
+  if(NvPmResultFails(mNvPmApi->RemoveAllCounters(mNvPmCtx), "call to 'NvPmApi::RemoveAllCounters'"))
+  {
+    return false;
+  }
+
+  mSelectedExternalIds.clear();
+  mSelectedInternalIds.clear();
+
+  std::for_each(counters.begin(), counters.end(), [
+                                                      &selExternalIds = mSelectedExternalIds,
+                                                      &selInternalIds = mSelectedInternalIds,
+                                                      &internalIds = mInternalIds
```

Comment:
Is there any reason you're using `std::for_each` and a callback here instead of just a simple for loop? You can use `for(GPUCounter counter : counters)` if you prefer that style here.